pinot-common/src/main/java/com/linkedin/pinot/common/utils/MmapUtils.java

```diff
@@ -58,16 +59,71 @@
   }
 
   public static class AllocationContext {
+    private final String fileName;
+    private final String parentFileName;
+    private final String parentPathName;
+
     private final String context;
     private final AllocationType allocationType;
 
-    public AllocationContext(String context, AllocationType allocationType) {
+    public static final Joiner PATH_JOINER = Joiner.on(File.separatorChar).skipNulls();
+
+    AllocationContext(String context, AllocationType allocationType) {
       this.context = context;
       this.allocationType = allocationType;
+      this.fileName = null;
+      this.parentFileName = null;
+      this.parentPathName = null;
+    }
+
+    AllocationContext(File file, String details, AllocationType allocationType) {
+      context = details;
+      this.allocationType = allocationType;
+
+      // We separate the path into three components to lower the overall on-heap memory usage when there are lots of
+      // redundant paths. Paths that are memory mapped are usually <storage directory>/<segment name>/<column name>.ext
```

Comment:
Yes, I looked at the server that ran out of memory this morning and it had 55k+ AllocationContexts and it also had the merged segment format. While this is a small amount of the heap (~35MiB), this should save some space.

One current `AllocationContext` has a size of 32 bytes, this will roughly double its size to 64 bytes, but will enable the sharing of String instances between multiple allocation contexts, so overall usage should be lower.

Most of the heap was still taken by other Strings, but that's still an optimization. 😄